marshalable/src/lib.rs

```diff
@@ -34,6 +34,49 @@ pub trait Marshalable {
     fn try_marshal(&self, buffer: &mut [u8]) -> TpmRcResult<usize>;
 }
 
+/// Defines the ability to marshal an enum by its variant data alone.
+///
+/// The separation of this trait from `Marshalable` is useful for cases where
+/// an enum's data is not directly after the variant's selector. This is
+/// something that can happen in some TPM types, so we need to be able to
+/// support it (and it needs to be able to function cross-crates in order to
+/// support vendors creating their own types, commands, etc).
+///
+/// # Notice
+///
+/// Today, you can get this trait simply by deriving the `Marshalable` trait.
+/// However, in the future we will be separating this out into a different
+/// derive proc-macro.
+///
+/// See: https://github.com/tpm-rs/tpm-rs/issues/84
+pub trait MarshalableVariant {
+    /// The type responsible for holding the discriminant in the enum.
+    type Selector: Copy;
+
+    /// Returns the discriminant of the current enum.
+    fn discriminant(&self) -> Self::Selector;
+
+    /// Tries to unmarshal into an enum with a specific selector's data.
+    ///
+    /// Because the way the data in `buffer` is interpreted changes depending on
+    /// the variant we are unmarshaling, we have to be told explicitly which
+    /// variant is being targeted.
+    fn try_unmarshal_variant(
+        selector: Self::Selector,
+        buffer: &mut UnmarshalBuf,
+    ) -> TpmRcResult<Self>
+    where
+        Self: Sized;
```

Comment:
Optional Nit: `Where Self: Sized` is the same as ` pub trait MarshalableVariant: Sized`  But the later looks simpler.

Reply:
I suppose since `MarshalableVariant` deals specifically with enums, one could make the argument for `MarshalableVariant` having a supertrait of `Sized` even if `Marshalable` doesn't...

Anyways, probably best to discuss this in a separate PR, I'll send one out after merging. Thanks!